cw_monero/lib/api/transaction_history.dart

```diff
@@ -16,20 +18,59 @@ String getTxKey(String txId) {
 
 monero.TransactionHistory? txhistory;
 
-void refreshTransactions() {
+bool isRefreshingTx = false;
+
+Future<void> refreshTransactions() async {
+  if (isRefreshingTx == true) return;
+  await txHistoryMutex.acquire();
+  isRefreshingTx = true;
   txhistory ??= monero.Wallet_history(wptr!);
-  monero.TransactionHistory_refresh(txhistory!);
+  final ptr = txhistory!.address;
+  await Isolate.run(() {
+    monero.TransactionHistory_refresh(Pointer.fromAddress(ptr));
+  });
+  isRefreshingTx = false;
+  txHistoryMutex.release();
 }
 
 int countOfTransactions() => monero.TransactionHistory_count(txhistory!);
 
-List<Transaction> getAllTransactions() {
+List<Transaction> cachedAllTxList = [];
+int cachedForWptr = 0;
+final txHistoryMutex = Mutex();
+Future<List<Transaction>> getAllTransactions({bool enableDelay = false}) async {
   List<Transaction> dummyTxs = [];
 
+  await txHistoryMutex.acquire();
   txhistory ??= monero.Wallet_history(wptr!);
-  monero.TransactionHistory_refresh(txhistory!);
   int size = countOfTransactions();
-  final list = List.generate(size, (index) => Transaction(txInfo: monero.TransactionHistory_transaction(txhistory!, index: index)));
+  final Stopwatch? stopwatch = Stopwatch()..start();
+  final List<Transaction> list = [];
+  if (cachedAllTxList.length == size && cachedForWptr == wptr!.address) {
+    list.addAll(cachedAllTxList);
+    monero.debugCallLength["CW_Transaction_list_construct_cached"] ??= <int>[];
+    monero.debugCallLength["CW_Transaction_list_construct_cached"]!.add(stopwatch?.elapsedMicroseconds??0);
+    txHistoryMutex.release();
+  } else {
+    final List<Transaction> newList = [];
+    for (var i = 0; i < size; i++) {
+      newList.add(
+        Transaction(txInfo: monero.TransactionHistory_transaction(txhistory!, index: i))
+      );
+      if (enableDelay) {
+        await Future.delayed(Duration.zero);
```

Comment:
What is the point of this Future?

Reply:
to give the UI some time to render, future is being put at the end of the scheduler queue (reusing my JS knowledge here, but it works)